### PR TITLE
Use Mutex to prevent multiple prompts

### PR DESCRIPTION
**Pull request by goagain**
# Background
Some users encounter that when accessing a resource at the same time, if the token has not been cached, many authentication dialog boxes will pop up, which is quite annoying.

Ref: #3 may contains more context.
# Solution
To avoid multiple prompts, a named mutex is used, which can prevent in both single process and multiple processes.
See [Mutex: Local vs Global](https://docs.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=net-6.0#remarks)

# Testcase
Run the following command parallel but in one terminal server sessions.
dotnet run --project ./src/AzureAuth --debug --resource={resource_id} --client={client_id} --tenant={tenant_id} -m web

After discussing with Kyle and Tucker, we can simplify by literally using Mutex. Since unit test on parallel scenario is difficult, we can test manually.

### CODE REVIEW

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:267]: Don't forget to mention the `this.Client` as well.
```suggestion
                // When running multiple AzureAuth processes with the same resource, client, and tenant IDs,

```

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:273]: This is just my opinion, but it might be a little cleaner to use string interpolation here. Also, consider using an underscore to separate each of these IDs in case the lock name needs to be read for debugging later.
```suggestion
                string lockName = $"Global\\{this.Resource}_{this.Client}_{this.Tenant}";
```

Comment (reillysiemens) [reply src/AzureAuth/CommandMain.cs:273]: Yet another consideration would be to hash (maybe with a SHA algorithm?) all three values so that they aren't accidentally leaked anywhere else in the OS (I don't actually know what C# does with the lock name). I'm not sure if that's necessary as I don't know if this trio is really considered "secret", but it might be worth thinking about.

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:289]: I have a small grammatical suggestion here and I think just a little extra context would be useful as well.
```suggestion
                        this.logger.LogWarning("The authentication attempt mutex was abandoned. Another thread or process may have exited unexpectedly.");
```

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:285]: There's a small grammatical correction that can be made here.
```suggestion
                    // An AbandonedMutexException could be thrown if another process exits without releasing the mutex correctly.

```

**Review by reillysiemens — COMMENTED** [review]
@goagain: This looks like a much more concise approach than #3! Really quick turnaround too after our last discussion.

I think it will be good to have other team members review this as well, so everyone understands this piece of the code, but I would be happy to approve this after you make some minor tweaks. 😀

**Comment by goagain** [comment]
> @goagain: This looks like a much more concise approach than #3! Really quick turnaround too after our last discussion.
> 
> I think it will be good to have other team members review this as well, so everyone understands this piece of the code, but I would be happy to approve this after you make some minor tweaks. 😀

Thank you. So many grammatical error : (

**Comment by kyle-rader** [comment]
> > @goagain: This looks like a much more concise approach than #3! Really quick turnaround too after our last discussion.
> > I think it will be good to have other team members review this as well, so everyone understands this piece of the code, but I would be happy to approve this after you make some minor tweaks. 😀
> 
> Thank you. So many grammatical error : (

It's okay! English is a difficult language to begin with and we all make mistakes here. We're all here to help one another learn and improve through practice :)

**Comment by reillysiemens** [comment]
@goagain: Honestly, I didn't think there were that many grammatical errors. I promise you I'd have more in any language that wasn't English. 😆

**Review by kyle-rader — APPROVED** [review]

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:286]: I think this comment could use just a little more clarification, but other than that things look good to me.

**Review by reillysiemens — APPROVED** [review]

File: src/AzureAuth/CommandMain.cs
Comment (goagain) [reply src/AzureAuth/CommandMain.cs]: I have the comment in this code block more detailed. Is that work for you?

File: src/AzureAuth/CommandMain.cs
Comment (mjcheetham) [inline src/AzureAuth/CommandMain.cs:271]: According to the `Mutex` documentation, using the `Global\` (or `Local\`) prefix for the name only applies to running on a Terminal Services server. From the [Remarks section of the docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=net-6.0#remarks):

> Note: On a server that is running Terminal Services, a named system mutex can have two levels of visibility. If its name begins with the prefix "Global\\", the mutex is visible in all terminal server sessions. If its name begins with the prefix "Local\\", the mutex is visible only in the terminal server session where it was created. In that case, a separate mutex with the same name can exist in each of the other terminal server sessions on the server. If you do not specify a prefix when you create a named mutex, it takes the prefix "Local\\". Within a terminal server session, two mutexes whose names differ only by their prefixes are separate mutexes, and both are visible to all processes in the terminal server session. That is, the prefix names "Global\\" and "Local\\" describe the scope of the mutex name relative to terminal server sessions, not relative to processes.

Not sure if you intended to restrict to only one simultaneous global auth prompt across multiple concurrent users.

**Review by mjcheetham — COMMENTED** [review]
Hello! 👋 I work on the Git Credential Manager team that also uses MSAL in a CLI tool. I'm excited to see the `authcli` hit public release! 🎉 😁 

I hate to jump in uninvited to comment here, but I had too considered adding some mutex in GCM to prevent multiple auth prompts in the past. I never actually got around to doing so, but I remember in my exploration reading about the multiple-user one-machine scenarios as commented below. Things may have changed in the implementation since then (over a year ago) so feel free to ignore me of course 😅

File: src/AzureAuth/CommandMain.cs
Comment (goagain) [reply src/AzureAuth/CommandMain.cs:271]: Hi @mjcheetham , thank you for your comment. Overall, yes, we do want restrict to only one auth prompt. For a same resource combination **{this.Resource}\_{this.Client}\_{this.Tenant}**, only the first successfully logged in window takes effect, the rest will use the cached authorization token.

Comment (kyle-rader) [reply src/AzureAuth/CommandMain.cs:271]: @mjcheetham  thanks so much for jumping in! Please do not hesitate, you are welcome here :). I agree we actually want `LOCAL\` (which if left off is the default) because of the scenario where two different logged in users are authenticating with the same triplet of GUIDs. 

It turns out the behavior of LOCAL, on Mac, also means two login shells actually don't get mutexed when they should,. So we will continue testing and might need to make this choice platform specific.

Comment (reillysiemens) [reply src/AzureAuth/CommandMain.cs:271]: @mjcheetham: Your uninvited commentary is actually a huge win for us. This kind of open, community-driven feedback is half the benefit of moving to GitHub for us. Thank you! ❤️

File: src/AzureAuth/CommandMain.cs
Comment (goagain) [reply src/AzureAuth/CommandMain.cs:280]: It will keep waiting until the mutex released.

Comment (shalinikhare27) [reply src/AzureAuth/CommandMain.cs:280]: Is that going to cause any experience issues?

Comment (kyle-rader) [reply src/AzureAuth/CommandMain.cs:280]: We still have a separate bug to enable a reliable timeout. So that _will_ get addressed :).

File: src/AzureAuth/CommandMain.cs
Comment (goagain) [reply src/AzureAuth/CommandMain.cs:271]: I wonder the definition of a **session**, if a client invoke our CLI multiple times, are these CLI actually in one **session**? If so, I guess local mutex is enough for this case.

Comment (goagain) [reply src/AzureAuth/CommandMain.cs:276]: @reillysiemens Shalini and I tested on MacOS. What we got at the time was that `Local` behaved differently on Windows and MacOS. On Windows, different PowerShells can share a local mutex, but on MacOS, multiple shells open multiple authentication prompts. If I use `Global`, the mutex works on the OS layer.

File: src/AzureAuth/CommandMain.cs
Comment (goagain) [reply src/AzureAuth/CommandMain.cs]: @reillysiemens I added a 15-minute timeout for acquiring the mutex. If timeout, a `TimeoutException` will be thrown, which can be catched by telemetry. Does that work for you?

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:56]:
```suggestion
        /// The maximum time we will wait to acquire a mutex around prompting the user.

```

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:58]:
```suggestion
        private TimeSpan promptMutexTimeout = TimeSpan.FromMinutes(15);

```
Might be worth a little more specificity here. This particular mutex is guarding the number of user prompts.

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [inline src/AzureAuth/CommandMain.cs:273]: Looks like there's an accidental extra newline here. I think you can just delete this line.

**Review by reillysiemens — DISMISSED** [review]
@goagain, thanks so much for the hard work on this feature! And again, thanks for your patience. This has been back and forth many times.

I've made some small suggestions here for final changes, but assuming those are fine with you I'm very happy to see this merged.

**Ship It!** 🚀🚢

**Review by shalinikhare27 — DISMISSED** [review]

**Comment by goagain** [comment]
@reillysiemens @shalinikhare27 Thanks for suggestion and reviewing. Seems like you need to re-approve again after applying suggestions.

**Review by reillysiemens — APPROVED** [review]

**Review by shalinikhare27 — APPROVED** [review]

File: src/AzureAuth/CommandMain.cs
Comment (reillysiemens) [reply src/AzureAuth/CommandMain.cs]: Just to close the loop here, @mjcheetham, the final decision we landed on here was to use `Local` on all supported platforms. As you pointed out it definitely seems like the right call on Windows and has essentially the same behavior on macOS except that different login shells (and therefore different terminal windows) can prompt at the same time. Likelihood of two uncached calls from two different terminals happening at the same time is slim though, so we figured that's a reasonable risk to take on for now.

**Pull request merged**
